### ISSUE TITLE
Support pytest --pdb on test failure to persist mocked s3 object access

### ISSUE THREAD

**Issue by dazza-codes** [spulec/moto#2066]
Similar to #1560 although this issue is about dropping into pdb using the pytest `--pdb` option.

Versions:
```
$ pip freeze | grep -i -E 'boto|moto|pytest'
boto==2.49.0
boto3==1.9.96
botocore==1.12.96
moto==1.3.7
pytest==4.2.1
pytest-benchmark==3.2.2
pytest-cov==2.6.1
pytest-datadir==1.3.0
pytest-forked==1.0.2
pytest-mock==1.10.1
pytest-voluptuous==1.1.0
pytest-xdist==1.26.1
```

```python
# snippet from tests/test_moto_pdb_interaction.py
@mock_s3
def test_pdb_interaction():
    bucket_name = 'test_bucket'
    key = 'test_key/test_file.txt'
    # mock an s3 object
    conn = boto3.resource('s3', region_name='us-east-1')
    bucket = conn.create_bucket(Bucket=bucket_name)
    bucket.put_object(Key=key, Body='test data\n')
    assert False  # drop into pdb here
```

```
$ pytest -n 0 --pdb tests/test_moto_pdb_interaction.py
-> assert False
(Pdb) bucket
s3.Bucket(name='test_bucket')
(Pdb) conn.meta.client.head_object(Bucket=bucket_name, Key=key)
*** botocore.exceptions.ClientError: An error occurred (403) when calling the HeadObject operation: Forbidden
```

**Comment by spulec** [comment]
Sorry, I'm a bit confused here. Is the example at the end showing what happens or what you want? If the former, I don't think I understand the issue since it seems to work.

**Comment by spulec** [comment]
After looking at it again, I think I understand now. When it goes to pdb mode, we lose the context.

Do you know if other mocking libraries (like mock) are able to persist even in pdb mode?

**Comment by darrenleeweber** [comment]
> After looking at it again, I think I understand now. When it goes to pdb mode, we lose the context.
- right

> Do you know if other mocking libraries (like mock) are able to persist even in pdb mode?
- good question, if it's not specific to moto, then punt on it

**Comment by darrenleeweber** [comment]
A small test using the `requests-mock` library works with pdb, e.g.
```
    def test_epsg_lookup_success(wkt_match, epsg_match_response):
        with requests_mock.mock() as mock:
            mock.get('http://prj2epsg.org/search.json', text=json.dumps(epsg_match_response))
    
>           assert False
E           assert False

tests/test_crs.py:40: AssertionError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /home/joe/src/org/gis/tests/test_crs.py(40)test_epsg_lookup_success()
-> assert False
(Pdb) epsg_code = crs.wkt_to_epsg_lookup(wkt_match)
(Pdb) epsg_code
{'code': '2278', 'name': 'NAD_1983_StatePlane_Texas_South_Central_FIPS_4204_Feet', 'url': 'http://prj2epsg.org/epsg/2278.json'}
```
No idea whether this is related to hitting a breakpoint within a context manager, but at least it seems to work.

**Comment by bblommers** [comment]
From what I can see, `pdb` will always exit the context manager, if it encounters an error.

Using a simple test case:
```
def test_epsg_lookup_success():
    with requests_mock.mock() as mock:
        hi = "there"
        mock.get('http://prj2epsg.org/search.json', text="response")

        breakpoint()
```
PDB hits the breakpoint that's still inside the decorator. So any calls to `requests` will be caught:
```
pytest -sv --pdb tests/test_pdb.py 

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB set_trace >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

(Pdb) requests.get("http://www.google.com")
*** requests_mock.exceptions.NoMockAddress: No mock address: GET http://www.google.com/
(Pdb) requests.get("http://prj2epsg.org/search.json")
<Response [200]>
```
-------------
Using that same example, but throwing an error instead of a breakpoint:
```
def test_epsg_lookup_success():
    with requests_mock.mock() as mock:
        hi = "there"
        mock.get('http://prj2epsg.org/search.json', text="response")

        assert False
```
Now PDB exits the contextmanager, so that calls to requests are no longer intercepted:
```
(Pdb) requests.get("http://www.google.com")
*** NameError: name 'requests' is not defined
(Pdb) import requests
(Pdb) requests.get("http://www.google.com")
<Response [200]>
```
-----------------
TLDR: Unless I'm missing something.. there is not really something we can do about this.
PDB will always exit the context when encountering an error, and we always need to un-mock boto3 when exiting the context.
